File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/757e23ba8cb9ccfc4c55dd5c690349dd.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>

#define MAX_N 2000
#define MAX_K 2000

int64_t max(int64_t a, int64_t b) {
    return (a > b) ? a : b;
}

void add_item(int64_t knapsack[], int size, int64_t sum, int K) {
    for (int k = K - size; k >= 0; k--) {
        knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum);
    }
}

void recurse(int start, int end, int N, int K, int64_t A[][MAX_K], int64_t A_sum[], int64_t knapsack[], int64_t *best) {
    if (start >= end) {
        return;
    }

    if (end - start == 1) {
        int64_t sum = 0;
        *best = max(*best, knapsack[K]);

        for (int prefix = 1; prefix <= A[start][0]; prefix++) {
            sum += A[start][prefix];
            *best = max(*best, sum + knapsack[K - prefix]);
        }

        return;
    }

    int mid = (start + end) / 2;
    int64_t state[MAX_K];
    for (int i = 0; i <= K; i++) {
        state[i] = knapsack[i];
    }

    for (int i = start; i < mid; i++) {
        add_item(state, A[i][0], A_sum[i], K);
    }

    recurse(mid, end, N, K, A, A_sum, state, best);

    for (int i = 0; i <= K; i++) {
        state[i] = knapsack[i];
    }

    for (int i = mid; i < end; i++) {
        add_item(state, A[i][0], A_sum[i], K);
    }

    recurse(start, mid, N, K, A, A_sum, state, best);
}

void solve() {
    int N, K;
    scanf("%d %d", &N, &K);

    int64_t A[MAX_N][MAX_K];
    int64_t A_sum[MAX_N];

    for (int i = 0; i < N; i++) {
        int t;
        scanf("%d", &t);

        for (int j = 0; j < t; j++) {
            scanf("%lld", &A[i][j]);
        }

        if (t > K) {
            t = K;
        }

        A_sum[i] = 0;
        for (int j = 0; j < t; j++) {
            A_sum[i] += A[i][j];
        }
    }

    int64_t best = 0;
    int64_t knapsack[MAX_K + 1] = {0};
    recurse(0, N, N, K, A, A_sum, knapsack, &best);

    printf("%lld\n", best);
}

int main() {
    solve();
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/757e23ba8cb9ccfc4c55dd5c690349dd.py
import sys

class DSum:
    class Solver:
        def __init__(self, in_stream, out_stream):
            self._in = in_stream
            self._out = out_stream

        def read(self, *args):
            for arg in args:
                arg[0] = int(self._in.readline().strip())

        def write(self, *args):
            for arg in args:
                self._out.write(str(arg) + '\n')

        def solve(self):
            N, K = 0, 0
            A = []
            A_sum = []
            best = 0

            self.read([N], [K])
            for _ in range(N):
                t = 0
                self.read([t])
                a = list(map(int, self._in.readline().strip().split()))
                A.append(a[:min(t, K)])
                A_sum.append(sum(A[-1]))

            knapsack = [0] * (K + 1)

            def add_item(knapsack, size, s):
                for k in range(K - size, -1, -1):
                    knapsack[k + size] = max(knapsack[k + size], knapsack[k] + s)

            def recurse(start, end, knapsack):
                nonlocal best

                if start >= end:
                    return

                if end - start == 1:
                    s = 0
                    best = max(best, knapsack[K])

                    for prefix in range(1, len(A[start]) + 1):
                        s += A[start][prefix - 1]
                        best = max(best, s + knapsack[K - prefix])

                    return

                mid = (start + end) // 2
                state = knapsack[:]

                for i in range(start, mid):
                    add_item(state, len(A[i]), A_sum[i])

                recurse(mid, end, state)
                state = knapsack[:]

                for i in range(mid, end):
                    add_item(state, len(A[i]), A_sum[i])

                recurse(start, mid, state)

            recurse(0, N, knapsack)
            self.write(best)

    @staticmethod
    def solve(in_stream, out_stream):
        DSum.Solver(in_stream, out_stream).solve()


if __name__ == '__main__':
    DSum.solve(sys.stdin, sys.stdout)
